### PR TITLE
Reduce allocations in Get-Content cmdlet

### PR DESCRIPTION
**Pull request by iSazonov**
## PR Summary

Current implementation do __huge__ extra allocations for every line of a file and at worst for every char (!). The fix exclude the extra allocations by moving initializations in a constructor and using Span<char> for temporary buffer.

The Boyer-Moore string search algorithm was optimized. Use simple mapping Unicode chars to byte. It works well for common text files with chars from one-two languages.

You could review commit by commit.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs [hunk 1]
```diff
@@ -222,50 +225,6 @@ internal class FileSystemContentReaderWriter : IContentReader, IContentWriter
...
-        /// <param name="waitForChanges">
-        /// If true, we will perform blocking reads on the file, waiting for new content to be appended
-        /// </param>
-        /// <param name="provider">
-        /// The CmdletProvider invoking this stream
-        /// </param>
-        /// <param name="isRawStream">
-        /// Indicates raw stream.
-        /// </param>
-        public FileSystemContentReaderWriter(
```
Comment (anmenaga) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: Why was this removed?

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs [hunk 2]
```diff
@@ -314,8 +273,29 @@ internal class FileSystemContentReaderWriter : IContentReader, IContentWriter
...
+                const int DefaultLineLength = 256;
+                _currentLineContent = new StringBuilder(DefaultLineLength);
+
+                // For Boyer-Moore string search algorithm.
+                // Populate the offset lookups.
+                // These will tell us the maximum number of characters
+                // we can read to generate another possible match.
+                // If we read more characters than this, we risk consuming
+                // more of the stream than we need.
+                _offsetDictionary = new Dictionary<char, int>(_delimiter.Length);
```
Comment (PaulHigin) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: This looks a little bit dangerous.  If _usingDelimiter is set to true anywhere else, _currentLineContent, _offsetDictionary will be null during later reads.  I feel we should at least have an assert where these are used to ensure they are not null and as a warning if the code somehow gets changed.

**Review by PaulHigin — CHANGES_REQUESTED** [review]
Overall this looks good to me.  I agree that the offset Dictionary could be a perf issue for simple delimiter characters.  Have you performed any tests to verify perf gains?  I imagine they would be significant, in some case, by switching to use Span<T>.

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: It is unneeded and used __internal__ constructor and we can safely remove it. (Otherwise we have to fix it to follow our changes in another constructor with delimiter parameter.)

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: If somebody make the wrong change our tests will be crashed with null reference exception and we'll never merge the code. I don't know where we could set the assert and how it can help more then the null reference exception.

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs [hunk 3]
```diff
@@ -578,103 +558,112 @@ private bool ReadByLine(bool waitChanges, ArrayList blocks, bool readBackward)
...
+            {
+                // when -Raw is used we want to anyway read the whole thing
+                // so avoiding the while loop by reading the entire content.
+                string contentRead = _reader.ReadToEnd();
+                if (contentRead.Length > 0)
+                {
+                    blocks.Add(contentRead);
+                }
+
+                return _reader.Peek() != -1;
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: Fixed.

**Comment by iSazonov** [comment]
> I agree that the offset Dictionary could be a perf issue for simple delimiter characters. 

I don't found how quickly fix this. Will investigate (out of the PR).

> Have you performed any tests to verify perf gains?

No. I accidentally discovered this by reading code. Once I realized that we allocate in heap for every char in a worst case and for every line I did not see the point to run PerfView.

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs [hunk 4]
```diff
@@ -287,13 +288,34 @@ internal class FileSystemContentReaderWriter : IContentReader, IContentWriter
...
+                }
+
+                // If next char from file is in search pattern we should calculate a safe shift.
+                char currentChar;
+                byte lowByte;
+                for (var i = 0 ; i < _delimiter.Length; i++)
+                {
+                    currentChar = _delimiter[i];
+                    lowByte = Unsafe.As<char,byte>(ref currentChar);
+                    _offsetDictionary[lowByte] = _delimiter.Length - _delimiter.LastIndexOf(currentChar) - 1;
```
Comment (PaulHigin) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: I think this can be
```
_offsetDictionary[lowByte] = _delimiter.Length - i - 1;
```
For duplicate chars the last index will end up in _offsetDictionary[lowByte]

Also, do we need to be worried about char collisions using the low byte?

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs [hunk 5]
```diff
@@ -585,6 +607,7 @@ private bool ReadDelimited(bool waitChanges, ArrayList blocks, bool readBackward
             Span<char> readBuffer = stackalloc char[currentOffset];
             bool delimiterNotFound = true;
             _currentLineContent.Clear();
+            char currentChar;
```
Comment (PaulHigin) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: `currentChar` can be defined below where it is used.

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs (hunk 4 above)
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: > Also, do we need to be worried about char collisions using the low byte?

There is a comment in the code. If we have a file with chars from many languages we'll have many collisions and lost performance. In practice, we can expect that files contains chars from one-two languages. In the common case we'll have no collisions and best performance.

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: > I think this can be

Thanks! Good catch!

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs (hunk 5 above)
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: Fixed.

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by iSazonov** [comment]
@anmenaga  @SteveL-MSFT Have you any comments?

**Comment by anmenaga** [comment]
`Get-Content` is a frequently used cmdlet. I would like to see a green `[feature]` test run for this. Thanks.
